src/main/kotlin/de/nordgedanken/auto_hotkey/ide/documentation/AhkDocumentationProvider.kt

```diff
@@ -0,0 +1,150 @@
+ * AutoHotkey.chm help file present in the AutoHotkey home directory
+ */
+class AhkDocumentationProvider : DocumentationProvider, ExternalDocumentationHandler {
+
+    override fun getCustomDocumentationElement(
+        editor: Editor,
+        file: PsiFile,
+        contextElement: PsiElement?,
+        targetOffset: Int
+    ): PsiElement? {
+        return contextElement
+    }
+
+    override fun getUrlFor(element: PsiElement?, originalElement: PsiElement?): MutableList<String>? {
+        val url = getUrl(element) ?: return null
+        return mutableListOf(url)
+    }
+
+    private fun getUrl(element: PsiElement?): String? {
+        element ?: return null
+
+        val chm = try {
+            ChmArchiveUtil.getChmArchiveHandler(element.project)
+        } catch (e: IllegalStateException) {
+            return null
+        }
+
+        if (element.text.startsWith("A_"))
+            return "https://www.autohotkey.com/docs/Variables.htm#" + element.text.drop(2)
+
+        val pathInChm = ChmArchiveUtil.getPathInChm(chm, element.text)
+            ?: return null
+
+        return "https://www.autohotkey.com/$pathInChm"
+    }
+
+    override fun generateDoc(element: PsiElement?, originalElement: PsiElement?): String? {
+        if (element !is LeafPsiElement) return null
+        if (element.elementType !is AhkTokenType) return null
+
+        val chm = try {
+            ChmArchiveUtil.getChmArchiveHandler(element.project)
+        } catch (e: IllegalStateException) {
+            return e.message
+        }
+
+        if (element.text.startsWith("A_"))
+            return extractVariableText(chm, element.text)
+
+        val pathInChm = ChmArchiveUtil.getPathInChm(chm, element.text)
+            ?: return null
+
+        return extractHtmlText(chm, pathInChm)
+    }
+
+    private fun extractVariableText(chm: ArchiveHandler, variableName: String): String? {
+        val variablesPage = ChmArchiveUtil.getPathInChm(chm, "Variables")
+            ?: return null
+
+        return try {
+            String(chm.contentsToByteArray(variablesPage))
+                .substringAfter("<td>$variableName</td>")
+                .substringBefore("</tr>")
+                .replace("<td>", "<p>")
+                .replace("</td>", "</p>")
+                // disable anchor links
+                .replace("<a href=\"#", "<span href=\"")
+        } catch (e: FileNotFoundException) {
+            "File not found"
+        } catch (e: FileTooBigException) {
+            "File too big"
+        }
+
+    }
+
+    private fun extractHtmlText(
+        chm: ArchiveHandler,
+        pathInChm: String
+    ): String {
+        return try {
+            String(chm.contentsToByteArray(pathInChm))
+                // disable anchor links
+                .replace("<a href=\"#", "<span href=\"")
+        } catch (e: FileNotFoundException) {
+            "File not found"
+        } catch (e: FileTooBigException) {
+            "File too big"
+        }
```

Comment:
It seems the exception messages here are just a reflection of the exception name itself. It'll be simpler for us to apply kotlin's `runCatching` method here instead:

```kotlin
private fun extractHtmlText(chm: ArchiveHandler, pathInChm: String) = runCatching {
    String(chm.contentsToByteArray(pathInChm))
        // disable anchor links
        .replace("<a href=\"#", "<span href=\"")
}.onFailure { it.message }.getOrThrow()
```